### PR TITLE
Demo a lora API

### PR DESCRIPTION
**Pull request by mattdangerw**
Not sure this is fully correct yet, mainly want to get thoughts on the overall usage for now.

After chatting with Francois, I believe a method that wrapped a backbone without changing the original backbone whatsoever, would be fairly complex to write, particularly with the dual `tf.keras` and `keras-core` support.

It will be much easier (and potentially a simpler in usage), to do this in place. For that, we can use the approach Abheesht piloted in his guide.

The flow given here would be:
```python
classifier = keras_nlp.models.BertClassifier.from_preset(
    "bert_base_en_uncased",
    num_classes=2,
)
classifier.backbone.add_lora_layers()
classifier.summary()
classifier.fit(dataset)
classifier.backbone.merge_lora_layers()
classifier.evaluate(imdb_test)

gpt2_lm = keras_nlp.models.GPT2CausalLM.from_preset("gpt2_base_en")
gpt2_lm.backbone.add_lora_layers()
gpt2_lm.summary()
gpt2_lm.fit(dataset)
gpt2_lm.backbone.merge_lora_layers()
gpt2_lm.generate(["that's weird", "that's even weirder"], max_length=32)
```

### CODE REVIEW

File: keras_nlp/layers/modeling/lora_dense.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+from keras_nlp.backend import ops
+
+
+@keras_nlp_export("keras_nlp.layers.LoraDense")
+class LoraDense(keras.layers.Layer):
+    """A LoRA adapter layer for a dense input layer."""
+
+    def __init__(
+        self,
+        layer,
```
Comment (jbischof) [inline keras_nlp/layers/modeling/lora_dense.py:27]: input_layer?

File: keras_nlp/layers/modeling/lora_dense.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        if layer.built:
+            self.build((None, None, self.layer.kernel.shape[0]))
+
+    def build(self, inputs_shape):
+        self.lora_a.build(inputs_shape)
+        self.lora_b.build(self.lora_a.compute_output_shape(inputs_shape))
+        self.built = True
+
+    def merge_weights(self):
+        updates = ops.squeeze(
```
Comment (jbischof) [inline keras_nlp/layers/modeling/lora_dense.py:80]: Explanatory comment would be helpful. Why is this needed? When is this called?

File: keras_nlp/models/backbone.py [hunk 3]
```diff
@@ -48,12 +52,70 @@ def token_embedding(self, value):
...
+    @property
+    def lora_layer_paths(self):
+        """Path to model layers to replace with lora adapaters.
+
+        Should be overridden by subclasses who store key/value paths under
+        different names.
+        """
+        return ["_query_dense", "_value_dense"]
+
+    def add_lora_layers(self, rank=8, alpha=32):
```
Comment (jbischof) [inline keras_nlp/models/backbone.py:64]: I assume the user will be able to supply their own paths in the full version?

**Review by jbischof — COMMENTED** [review]
Thanks! Some initial questions as I found the `merge_weights` method confusing

File: keras_nlp/layers/modeling/lora_dense.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/lora_dense.py:27]: Yeah, I had played with `inner_layer` or `inner_dense`, but then I switched to this after discovering keras exports a wrapper layer https://www.tensorflow.org/api_docs/python/tf/keras/layers/Wrapper

That uses the argument `layer` and `layer.layer` as access to the inner.

File: keras_nlp/layers/modeling/lora_dense.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/lora_dense.py:80]: Yeah, will drop some docstrings and maybe rename this, could use improvement.

File: keras_nlp/models/backbone.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/models/backbone.py:64]: Backbone would be able to specify their own paths via the `lora_layer_paths` property.

But why would a user be specifying their own paths, don't really understand the use case. A highly motivated user could subclass a backbone and override `lora_layer_paths`, but I don't think that's something we would really document.

What's the use case for user supplied paths?

Comment (jbischof) [reply keras_nlp/models/backbone.py:64]: Just to allow 100% configurability. I noticed that the HF LoRA config allows the users to choose which "modules" are targeted ([link](https://huggingface.co/docs/peft/task_guides/semantic-similarity-lora)).

However, a brief glance across their examples does not reveal a common language for referencing "modules" (e.g., `"q"` or `"query"`), so I'm not sure how the user would know how to leverage this.

Comment (mattdangerw) [reply keras_nlp/models/backbone.py:64]: Potentially for 100% configurability you just to it yourself. E.g. find the precise layers you want to update, wrap then in a `LoraDense`, and proceed as normal. That is the truly flexible approach, and maybe 10-20 lines of code.

This current "find all layers by attribute name" is probably not robust to all possible cases, so probably best not to expose a user hook for it. We could try to make a robust mechanism, e.g. `*._self_attention._key_dense`, but first might be worth clarifying the user journeys we care about.

If we had a specific type configuration we want to support. E.g. tunable `bias` terms, lorifying the `key` dense, etc, we could definitely make sure to make out API supports that. But I am not really sure where those are yet.

Comment (jbischof) [reply keras_nlp/models/backbone.py:64]: So the current design is that for some models we will offer an in-place LoRA that modifies a prespecified list of layers?

I think this is a reasonable option but somehow I thought you had proposed something more ambitious when we first talked. I do think it would be possible to support an arbitrary regex and then try to find a matching LoRA class in the same way you are already doing in the "dense" layer. 

Potentially the two approaches could be combined so that we expose a low-level LoRA API that is pre-configured for most of our models using class properties.

I think the biggest challenge of the "no configuration" approach is that we could never change the choice of layers without breaking older versions of the package. Likely some safety valve is required.

Comment (mattdangerw) [reply keras_nlp/models/backbone.py:64]: > So the current design is that for some models we will offer an in-place LoRA that modifies a prespecified list of layers?

Yeah, this primarily would be LoRA that we works out of the box with all backbones w/ same code. You don't have to introspect your model to find the right regexes for query/value dense layers, we set that up for you.

This is what I think most users care about. Being able to go from model to model with the same code. "Lorifying" something custom (e.g. key projection), seems much more niche. Is it ever a good idea? (Genuinely not sure)

> I think this is a reasonable option but somehow I thought you had proposed something more ambitious when we first talked. I do think it would be possible to support an arbitrary regex and then try to find a matching LoRA class in the same way you are already doing in the "dense" layer.

Definitely possible! The question is whether that is actually a better experience then something like:

```python
# Custom LoRAfication of a backbone.
for i in range(backbone.num_layers):
    decoder_layer = backbone.get_layer("transformer_layer_{i}")
    decoder_layer._self_attention._query_dense = keras_nlp.layers.LoraDense(...)(
        decoder_layer._self_attention._query_dense
    )
    decoder_layer._self_attention._value_dense = keras_nlp.layers.LoraDense(...)(
        decoder_layer._self_attention._value_dense
    )
```

Building a fancier selection mechanism, perhaps based on layer name, might allow a user to say "lora only decoder self attention query/value projection in a seq2seq model". The pythonic flow above would allow for lora-ing any layers whatsoever, potentially with different configurations for each. Far more flexible.

I don't know the main axes that are actually worth playing with, so hard to asses between the two options. Just looking at the huggingface API, the most useful config option looks like [bias](https://huggingface.co/docs/peft/main/en/package_reference/tuners#peft.LoraConfig.bias) to me. But overall, I'm inclined to build something simple and go from there (always easier to add more configuration then take away).

> I think the biggest challenge of the "no configuration" approach is that we could never change the choice of layers without breaking older versions of the package. Likely some safety valve is required.

This I don't understand. If we expose a no configuration approach we leave ourselves more flexible. Say client code looks like:

```python
gpt2_lm.backbone.add_lora_layers()
```

If we release a new library version and checkpoints, the new library points to new checkpoints, the old library points to old checkpoints, we update our lora paths in the new library and "just works" everything works in all versions.

However, say client code looks like:

```python
gpt2_lm.backbone.add_lora_layers(layer_paths=["encoder.*._self_attention.query_dense"])
```

Then we are shit out of luck. If we update our checkpoints (say we want to rename `query_dense` to `query`), we can never do this without breaking client code, or coding a shim.

If we are worried about compatibility, exposing a minimal as possible API gives us much more flexibility I think. I am missing something?

Comment (mattdangerw) [reply keras_nlp/models/backbone.py:64]: Maybe one thing you are missing is that we can **never** update the checkpoint and old version of the library loads. Like core Keras, we have checkpoint sha-sums that ship with the library! So if we every want to push a new checkpoint, we have to release a new version on the library, and we have to release the checkpoint at a new URL location.
